### ISSUE TITLE
Windows 95 - 2000 no longer supported

### ISSUE THREAD

**Issue by rversteegen** [ohrrpgce/ohrrpgce#1241]
Support for Windows 95 through 2000 is broken in nightlies, even when using gfx_sdl builds (SDL2 requires Win XP+ anyway). This is due to a bug in the version of mingw-w64 used on the build machine. A couple people asked me to get it working again.

New builds needed
=================

Separately, there are no gfx_sdl builds available of stable releases, only of nightlies. I want to add these, to have support for Windows 95-2000, which we lost when switching to gfx_sdl2 as default in hróðvitnir. (I've given up on having a single Windows build that supports both old OSes and SDL2, though it is possible, using runtime linking. I just wish the default game.exe could show an error directing people to download the build for older OSes.)

Additionally, I want to increase the CPU requirement to Pentium 4+ by using SSE2, except in the build for old Windows OSes.

I'd also like to add an option to the Distribute Game menu to target these older Windows platforms, just like we have options to target obsolete 32-bit Mac and Linux. And then move those three obsolete targets to a separate section.

mingw-w64
=========

The build machine uses `docker/ohrrpgce-build-env-wine/Dockerfile`, which downloads the build linked from mingw-w64.org, which is surprisingly still GCC 8.1.0/mingw-w64 6.0.0, a few years out of date now.

Nightly builds don't run on Windows 2000, with the error "linked to missing export MSVCRT.DLL:___mb_cur_max_func". This function is present only in Win XP+.
I found that this is a mingw-w64 6.0.0 bug, fixed in 7.0.0 (see https://github.com/mirror/mingw-w64/commit/ebf4bf9d2a9)
However it's a mystery to me what version of Windows mingw-w64 actually aims to support, because 7.0.0 makes another change that requires Win XP+, and for example msys2 (since certain versions including mingw-w64 7.0.0) now only supports Win7+.

Everything works on Win98 when compiled with an old mxe toolchain (mingw-w64 5.0.3, GCC 5.4.0, FB 1.06.0).

But if you compile with mingw-w64 7.0.0 there are other problems:

Compiling on Windows using winlibs-mingw-w64-i686-9.3.0-7.0.0-r3-sjlj (mingw-w64 7.0.0, GCC 9.3.0, FB 1.08.1), which is the toolchain officially used for and recommended for FB 1.08/1.09, I got a "libwinpthread-1.dll not found" error when running on Win98. I think this is a dependency added by libstdc++ (mingw-w64's libstdc++'s thread support can be compiled in either "posix" (meaning using emulated pthreads) or "win32" (meaning using native winapi threading directly) configuration).

I also tried cross-compiling using a mxe i686-w64-mingw32.static toolchain (mingw-w64 7.0.0, GCC 5.5.0), but that had similar problems. And I compiled the SDL_mixer.dll with that toolchain, and it also depends on libwinpthread-1.dll.

So I tried statically linking to winpthread (using -Bstatic). That removes the libwinpthread-1.dll dependency (and only increases size a little) but then the .exe depends on `AddVectoredExceptionHandler` in kernel32.dll (Win XP+ only), which is called from libwinpthreads. I changed SConscript link with gcc instead of g++ to try to fix this (r12985, 872c270adf92) but that wasn't a solution, just moving the error elsewhere. After a lot of experimentation there seems to be no way to avoid the ultimate dependency on `AddVectoredExceptionHandler` by changing build options if using these "posix" mingw-w64 7.0.0 toolchains.

So I've patched winpthreads (https://github.com/rversteegen/mxe/commit/aa87b55a7) in my mxe fork to remove this dependency, and it now works on Win98 again. (`AddVectoredExceptionHandler` is only used for the `pthread_setname_np` function which is just an unimportant debugging aid.) Note that I'm 19 months behind upstream mxe, still using mingw-w64 7.0.0 and GCC 5.5.0. I'd rather we use a newer GCC.

I'll recompile SDL_mixer.dll with the patched toolchain.

Solutions
---------

On the Windows nightly build machine we need to do one of:
* Switch to mxe (`docker/ohrrpgce-mxe-build-env/`) and use my mxe patch, but I think that Dockerfile isn't working yet? I'll also want to upgrade to a newer GCC first, probably trivial but needs testing.
  * I could set up a cron job to create/upload these builds instead myself, but that's not pipelined
* Use an older mingw-w64 5.x.x toolchain. Undesirable and there's a risk it'll be incompatible with the official FB build unless you compile FB too
* Stick to the existing mingw-w64 6.0.0 toolchain for now but try to fix the `__mb_cur_max_func` error using a weak symbol (copying https://github.com/mirror/mingw-w64/commit/ebf4bf9d2a9). But we'll want to upgrade the toolchain eventually

HSpeak
======

Also, there's another problem, hspeak.exe built with Euphoria 4.1.0 does not run on Win 98. It produces an error:
```
Assertion failed
be-alloc.c, Line 549
Expression: ((uintptr_t)p & 7) == 0
```
Compiling with 4.0.5 does work.

We could go back to Euphoria 4.0.5 on the Windows build machine. I don't know of any advantages of using 4.1.0 on Windows.
(4.1.0 added 64-bit OS ports.)

**Comment by bob-the-hamster** [comment]
Correction: the nightly build machine does not use dockerfiles at all yet.

The Linux dockerfiles work, I just never switched over to them.

I don't remember if the windows dockerfiles we're ever fully functional

Nightly builds are still done on VirtualBox VMs

**Comment by rversteegen** [comment]
Ah. But `docker/ohrrpgce-build-env-wine/Dockerfile` seems to download the specific toolchain that's actually used on the build machine, as far as I can tell.

Another two possible solutions:
* use a mingw.org toolchain instead
* If the build machine is upgraded to a prebuilt mingw-w64 7.0.0 toolchain I could probably provide a patched libwinpthread.a to link against

Also I guess we probably need to statically link to libwinpthread but I'm not sure. I think the switch from g++ to gcc may have had the effect of not statically linking game/custom.exe to pthread and instead depending on pthread symbols in SDL_mixer.dll (since I patched mxe to have SDL_mixer.dll export symbols from libraries it contains). I've seen errors like `"GAME.EXE is linked to missing export SDL_MIXER.DLL:pthread_getspecific"`

**Comment by rversteegen** [comment]
Also, all newer versions of mingw-w64 presumably have the same problem with `AddVectoredExceptionHandler`, because the code which calls that hasn't changed since.

**Comment by rversteegen** [comment]
Since James just downgraded to Euphoria 4.0.5 I've just tried out the latest nightly in a Win 95 (OSR2) VM and found we're not there yet:
* game/custom don't run: `linked to missing export KERNEL32.DLL:IsDebuggerPresent`
* hspeak.exe doesn't run: `linked to missing export KERNEL32.DLL:ConvertThreadToFiber`. (hspeak.exe from Fufluns fails with the same error)

After copying in SDL_mixer.dll from Fufluns game/custom work fine. I tried out the Distribute Game options (except itch.io) and everything works except that rcedit.exe can't be run on Win 95, and Innosetup won't install (says it needs Windows 4.10 or later, i.e. Win NT). zip_exec.exe works. All programs in support/ run except for CrashSender1401.exe

I also notice some ignorable errors in c_debug.txt:
```
Error: MoveFileEx(...\general.reld.tmp, ...\general.reld) failed: This function is only valid in Win32 mode.
...copied file instead
```
Also happens with ohrrpgce.itm.resize.tmp files, etc. during upgrade. I don't know what "Win32 mode" is (surprisingly found nothing with Google yet) but I'm mildly concerned we're not in it!
Fufluns prints no such errors because it was released before the `MoveFileEx` code (in r11521 / c79a10cc6).

**Comment by rversteegen** [comment]
Game/Custom
--------------

As well as `IsDebuggerPresent`, winpthreads also uses `SetProcessAffinityMask` & `TryEnterCriticalSection` which are missing in kernel32.dll, and the latter can't be removed.

Now I understand that the 'win32' threading support in mingw-w64 is the original one and supports all 32-bit Windows, while 'posix' threading (winpthreads) is a recent addition [which is necessary to support](https://stackoverflow.com/questions/17242516/mingw-w64-threads-posix-vs-win32/30390278#30390278) C++11 mutexes and threads, but has higher system requirements. (It's pulled in by libmodplug which is C++. BTW in our C++ code I used FB mutexes wrapped in a C++ class rather than C++11 <mutex>, see `mutex.hpp`) That's why mxe switched to posix threads by default in 2019. I compiled the previous SDL_mixer.dll that worked on Win95 with mxe back in 2017. Also, back then, official FreeBASIC 1.06 builds were built with a mingw-w64 toolchain that also used win32 threads, although Fufluns didn't use them.

mxe can be compiled with win32 rather than posix threads with `make MXE_TARGETS=i686-w64-mingw32.static.win32`.

It doesn't feel like a brilliant idea to mix libraries built with two toolchains that use different ABIs but we were already doing it anyway. It seems it would only be an issue if attempting to link together C++ code or code that passed around pthread objects compiled with different toolchains. (Note SDL_mixer itself doesn't use libpthread, it uses SDL's mutex functions, which are directly implemented on winapi. It should be fine to use multiple threading libraries in the same executable like that).

So I recompiled SDL_mixer.dll with a `i686-w64-mingw32.static.win32` mxe toolchain but found that on Win95 it froze when switching away from a MIDI track. That was caused by the fix for #1060 which [I backported](https://github.com/rversteegen/SDL_mixer/commit/c24dbf48cd5cee084a393014e3712c2de4d353f2) from SDL_mixer 2.6 to 1.2, so I've backed that out. Disappointing. I wonder whether the fix is wrong, or it needs changes for SDL_mixer 1.2, or SDL mutexes are broken, or it's caused by my custom MIDI synthesizer, ... 

hspeak
---------

As for hspeak not running, [Windows 9x Notes](https://rpg.hamsterrepublic.com/ohrrpgce/Windows_9x_Notes) on the wiki mentions that it runs on Win 98 and later Win 95 versions, but not the one I have (I misspoke earlier, I have "4.00.950 C" not OSR2). So I'll ignore that.
